core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveDistinctAggregationForGroupedAggregation.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.iterative.Lookup;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+
+import java.util.HashSet;
+
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+
+/**
+ * Remove the aggregation node introduced by Distinct operator for a query
+ * containing grouped aggregation operator.
+ */
+public class RemoveDistinctAggregationForGroupedAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(RemoveDistinctAggregationForGroupedAggregation::isDistinctOperator);
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        if (doesGroupedAggregationSatisfyCondition(lookup.resolve(aggregationNode.getSource()),
+                lookup, aggregationNode)) {
+            return Result.ofPlanNode(aggregationNode.getSource());
+        }
+        else {
+            return Result.empty();
+        }
+    }
+
+    private static boolean isDistinctOperator(AggregationNode node)
+    {
+        return node.producesDistinctRows() && node.getGroupingSetCount() == 1;
+    }
+
+    private static boolean doesGroupedAggregationSatisfyCondition(PlanNode node, Lookup lookup, AggregationNode distinctNode)
+    {
+        if (node instanceof AggregationNode aggregationNode) {
+            return aggregationNode.getGroupingSets().getGroupingSetCount() == 1 &&
+                    new HashSet<>(distinctNode.getGroupingKeys()).containsAll(aggregationNode.getGroupingSets().getGroupingKeys());
+        }
+        else if (node instanceof ProjectNode projectNode) {
+            return doesGroupedAggregationSatisfyCondition(lookup.resolve(projectNode.getSource()), lookup, distinctNode);
```

Comment:
Add comment about project nodes introducing new symbols for computed expressions, and therefore preserving distinctness between the distinct aggregation and the child aggregation so long as all of the child aggregation keys remain present (without transformation by the project) in the distinct aggregation grouping keys.

Reply:
Done